### PR TITLE
Added AIE Trace PS Kernels for trace on the VCK5000

### PR DESCRIPTION
**Pull request by nishraptor**
Added 2 PS kernels to support event trace on the VCK5000. The aie_trace_config kernel configures the runtime trace performance counters and trace stream, and the aie_trace_gmio kernel configures the GMIO offload of the trace stream.

### CODE REVIEW

File: src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h [hunk 1]
```diff
@@ -54,17 +56,69 @@ namespace built_in {
...
+      uint32_t event_value = 0;
+      uint32_t counter_value = 0;
+
+  };
+
+  struct TileTraceData
+  {   
+    public:
+      uint32_t packet_type = 0;
+      uint32_t start_event = 28;
```
Comment (rozumx) [inline src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h:74]: Comment on what the magic number represents would be nice.

File: src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h [hunk 2]
```diff
@@ -54,17 +56,69 @@ namespace built_in {
...
+      TileTraceData  memory_trace_config;
+   
+      TileData(uint32_t c, uint32_t r) : column(c), row(r) {}
+  };
+
+  struct OutputConfiguration
+  {
+    public:
+      uint16_t numTiles;
+      uint32_t numTileCoreTraceEvents[9] = {0};
```
Comment (rozumx) [inline src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h:99]:
```suggestion
      uint32_t numTileCoreTraceEvents[9] = {};
```

File: src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h [hunk 3]
```diff
@@ -54,17 +56,69 @@ namespace built_in {
...
+      uint32_t numTileMemoryTraceEvents[9] = {0};
```
Comment (rozumx) [inline src/runtime_src/xdp/profile/plugin/aie_trace_new/x86/aie_trace_kernel_config.h:100]:
```suggestion
      uint32_t numTileMemoryTraceEvents[9] = {};
```

**Review by rozumx — CHANGES_REQUESTED** [review]
Some minor changes requested.

Side Note
Thank you for taking the time to do a self-code review prior to submitting the pull-request.   It shows and helps the code reviewers focus on what matters.

File: src/runtime_src/core/edge/ps_kernels/profiling/aie_trace_config/aie_trace_config.cpp [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+      //     << "Required core module trace slots for aie trace : "  << required;
+      //xrt_core::message::send(severity_level::info, "XRT", msg.str());
+      return false;
+    }
+
+    // Core Module broadcasts. 2 events for starting/ending trace
+    available = stats.getNumRsc(loc, XAIE_CORE_MOD, XAIE_BCAST_CHANNEL_RSC);
+    required = config.memoryCrossEventsBase.size() + 2;
+    if (available < required) {
+      // TODO: Send detailed message back to host code
```
[the remark this answers is not in the document]
Comment (nishraptor) [reply src/runtime_src/core/edge/ps_kernels/profiling/aie_trace_config/aie_trace_config.cpp:106]: All the TODO's are work that we are going to implement in the next PR, where we send the string data back to the host. , we are just marking it currently as we want to preserve the messages that we had on the edge (VCK190) implementation.

File: src/runtime_src/core/edge/ps_kernels/profiling/aie_trace_config/aie_trace_config.cpp [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+          cfgTile.core_trace_config.broadcast_mask_east = coreToMemBcMask;
+        } else {
+          cfgTile.core_trace_config.broadcast_mask_west = coreToMemBcMask;
+        }
+        // Done update config file
+
+        memoryEvents.clear();
+        numTileMemoryTraceEvents[numTraceEvents]++;
+
+        // TODO: Send success message back to host
```
[the remark this answers is not in the document]
Comment (nishraptor) [reply src/runtime_src/core/edge/ps_kernels/profiling/aie_trace_config/aie_trace_config.cpp:511]: When we implement it, we will have all the messaging within a function, but don't know the functionality of the mechanism we are going to use. The next PR will pull it into a function and clean all of them up.

**Review by jvillarre — APPROVED** [review]

**Review by rozumx — APPROVED** [review]

**Pull request merged**